keyboards/planck/keymaps/purple-rw/keymap.c

```diff
@@ -0,0 +1,548 @@
+                layer_off(_PADFUNC);
+                layer_off(_MODS);
+                layer_off(_MOUSE);
+                layer_off(_MACRO);
+                layer_on(_PLOVER);
+                if (!eeconfig_is_enabled()) {
+                    eeconfig_init();
+                }
+                keymap_config.raw = eeconfig_read_keymap();
+                keymap_config.nkro = 1;
+                eeconfig_update_keymap(keymap_config.raw);
+            } else {
+                layer_off(_ADJST);
+            }
+            return false;
+        case EXT_PLV:
+            if (record->event.pressed) {
+                #ifdef AUDIO_ENABLE
+                    PLAY_SONG(plover_gb_song);
+                #endif
+                layer_off(_PLOVER);
+                layer_on(get_highest_layer(default_layer_state));
+            } else {
+                layer_off(_ADJST);
+            }
+            return false;
+        case DB_TOGG:   /* fall thru */
+        case AG_NORM:   /* fall thru */
+        case AG_SWAP:   /* fall thru */
+        case AU_ON:     /* fall thru */
+        case AU_OFF:
+            if (!record->event.pressed) {
+                layer_off(_ADJST);
+            }
+    }
+    return true;
+}
+
+bool muse_mode = false;
+uint8_t last_muse_note = 0;
+uint16_t muse_counter = 0;
+uint8_t muse_offset = 70;
+uint16_t muse_tempo = 50;
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (muse_mode) {
+        if (IS_LAYER_ON(_RAISE)) {
+            if (clockwise) {
+                muse_offset++;
+            } else {
+                muse_offset--;
+            }
+        } else {
+            if (clockwise) {
+                muse_tempo+=1;
+            } else {
+                muse_tempo-=1;
+            }
+        }
+    } else {
+        if (clockwise) {
+            #ifdef MOUSEKEY_ENABLE
+                tap_code(KC_MS_WH_DOWN);
+            #else
+                tap_code(KC_PGDN);
+            #endif
+        } else {
+            #ifdef MOUSEKEY_ENABLE
+                tap_code(KC_MS_WH_UP);
+            #else
+                tap_code(KC_PGUP);
+            #endif
+        }
+    }
+    return true;
+}
+
+bool dip_switch_update_user(uint8_t index, bool active) {
+    switch (index) {
+        case 0: {
+#ifdef AUDIO_ENABLE
+            static bool play_sound = false;
+#endif
+            if (active) {
+#ifdef AUDIO_ENABLE
+                if (play_sound) { PLAY_SONG(plover_song); }
+#endif
+                layer_on(_ADJST);
+            } else {
+#ifdef AUDIO_ENABLE
+                if (play_sound) { PLAY_SONG(plover_gb_song); }
+#endif
+                layer_off(_ADJST);
+            }
+#ifdef AUDIO_ENABLE
+            play_sound = true;
+#endif
+            break;
+        }
+        case 1:
+            if (active) {
+                muse_mode = true;
+            } else {
+                muse_mode = false;
+            }
+    }
+    return true;
+}
+
+void matrix_scan_user(void) {
```

Comment:
Im not familiar with audio code, but if this code lives at `matrix_scan` for the only purpose of it being run every cycle of QMK's mainloop, it can most likely be moved to `housekeeping_task_user` which is the appropriate place for "repetitive tasks" logic

Reply:
Correct.  Specifically, the muse code was added long before the housekeeping function was, so that's why it uses matrix scan.